### PR TITLE
Fixed issues with PK filter and implicit cast

### PR DESCRIPTION
**Pull request by matriv**
- Minor code improvements
- Remove minNodeVersion from symbol Optimizer
  `minNodeVersion` has been introduced with: #10076 to implement a logic that checked the minimum version of the nodes currently in a CrateDB cluster, and use it to skip applying any symbol optimization rules that were introduced after that minNodeVersion. Since this mechanism hasn't been actually used since it was introduced (CrateDB: `4.2.0`), remove it, since it makes it difficult to call the symbol Optimizer from other parts of the code.

- Fixed issues with PK filter and implicit cast
  Given the following situation:
  ```
  CREATE TABLE tbl(a string PRIMARY KEY);
  INSERT INTO tbl(a) VALUES ('t');
  REFRESH TABLE t;
  SELECT * FROM tbl WHERE a = true;
  ```

  Previously, when trying to filter on the PK column with equality
  operator, but using a value of a different type (in this example
  `boolean`), then no results where returned because the dockeys where
  built incorrently. `ExpressionAnalyzer` creates an implicit cast on the
  column `a`, to match the type of the value (see: #10076 which introduced
  this concept). So the query of the `WHERE` clause becomes:
  ```
  a::boolean = true
  ```
  Because of that the value added to the dockeys is `true`, instead of `t`.
  
  To resolve this, we need to call `optimizeCasts()` before extracting
  the `pkMatches` with the use of `EqualityExtractor`, so that the cast is
  moved to the value:
  ```
  a = true::'string'
  ```
  and the `EqualityExtractor` normalizes the value to `t` which then
  is added to the list of dockeys.
  
  Fixes: #15326

### CODE REVIEW

File: server/src/main/java/io/crate/analyze/where/EqualityExtractor.java [hunk 1]
```diff
@@ -407,8 +394,10 @@ public Symbol visitReference(Reference ref, Context context) {
             return super.visitReference(ref, context);
         }
 
+        @Override
         public Symbol visitFunction(Function function, Context context) {
-
+            // Normalize the function so any casts on literals are evaluated
+            function = context.normalize.apply(function);
```
Comment (seut) [inline server/src/main/java/io/crate/analyze/where/EqualityExtractor.java:400]: Why do we normalize functions only but not simply the whole query instead?

Comment (matriv) [reply server/src/main/java/io/crate/analyze/where/EqualityExtractor.java]: You're right, I moved the normalization here, due to some other issue I had, but it's no longer necessary, I moved it to normalize the whole query instead.

**Review by seut — APPROVED** [review]
👍 thx

**Review by mfussenegger — COMMENTED** [review]
If we normalize the query with the optimizer - are then the `unwrapReferenceFromCast` calls still needed?


Changes look good to me.

**Comment by matriv** [comment]
> If we normalize the query with the optimizer - are then the `unwrapReferenceFromCast` calls still needed?

I've tried that, but it breaks the explicit cast situations, for example: https://github.com/crate/crate/blob/master/server/src/test/java/io/crate/analyze/where/EqualityExtractorTest.java#L336

as in those cases `optimizeCasts()` doesn't swap.

**Pull request merged**